File: PythonDummy/StructureSolver.py
import os
import sys
import argparse
from mpi4py import MPI
 
# check if PRECICE_ROOT is defined
if not os.getenv('PRECICE_ROOT'):
   print "ERROR: PRECICE_ROOT not defined!"
   exit(1)

precice_root = os.getenv('PRECICE_ROOT')
precice_python_adapter_root = precice_root+"/src/precice/adapters/python"
sys.path.insert(0, precice_python_adapter_root)

import PySolverInterface
from PySolverInterface import *

print "Starting Structure Solver..."

parser = argparse.ArgumentParser()
parser.add_argument("configurationFileName", help="Name of the xml config file.", type=str)
parser.add_argument("N", help="Number of mesh elements, needs to be equal for fluid and structure solver.", type=int)
args = parser.parse_args()

configFileName = args.configurationFileName
N              = args.N

print "N: " + str(N)

solverName = "STRUCTURE";

print "Configure preCICE..."
interface = PySolverInterface(solverName, 0, 1)
interface.configure(configFileName)
print "preCICE configured..."

dimensions = interface.getDimensions();

pressure               = [0.0001] * (N+1) #Pressure
crossSectionLength     = [1.0] * (N+1) #Cross-section length

meshID = interface.getMeshID("Structure_Nodes")
crossSectionLengthID = interface.getDataID("CrossSectionLength", meshID)
pressureID = interface.getDataID("Pressure", meshID)

vertexIDs = [0.0]*(N+1)
grid = [None]*(dimensions*(N+1))

for i in range(0, N+1):
   for dim in range(0, dimensions):
      grid[i*dimensions + dim]= i*(1-dim)

t = 0

interface.setMeshVertices(meshID, N+1, grid, vertexIDs)

print "Structure: init precice..."
interface.initialize()

if (interface.isActionRequired(PyActionWriteInitialData())):
   interface.writeBlockScalarData(crossSectionLengthID, N+1, vertexIDs, crossSectionLength)
   interface.fulfilledAction(PyActionWriteInitialData())

interface.initializeData();

if (interface.isReadDataAvailable()):
   interface.readBlockScalarData(pressureID, N+1, vertexIDs, pressure)

while (interface.isCouplingOngoing()):
   # When an implicit coupling scheme is used, checkpointing is required
   if (interface.isActionRequired(PyActionWriteIterationCheckpoint())):
      interface.fulfilledAction(PyActionWriteIterationCheckpoint())

   for i in range(0, N+1):
      crossSectionLength[i] = 4.0 / ((2.0 - pressure[i]) * (2.0 - pressure[i]))

   interface.writeBlockScalarData(pressureID, N+1, vertexIDs, pressure)
   interface.advance(0.01)
   interface.readBlockScalarData(crossSectionLengthID, N+1, vertexIDs, crossSectionLength)

   if (interface.isActionRequired(PyActionReadIterationCheckpoint())): # i.e. not yet converged
      interface.fulfilledAction(PyActionReadIterationCheckpoint())
   else:
      t = t + 1

print "Exiting StructureSolver"

interface.finalize()

File: cxx/StructureSolver_Parallel/StructureSolver.cpp
#include "StructureSolver.h"
#include "precice/SolverInterface.hpp"

#include <iostream>
#include <mpi.h>

using namespace precice;
using namespace precice::constants;

int main(int argc, char** argv)
{
  std::cout << "Starting Structure Solver..." << std::endl;
  if (argc != 3) {
    std::cout << std::endl;
    std::cout << "Structure: Usage: mpiexec -np <#procs> " << argv[0] << " <configurationFileName> <N>" << std::endl;
    std::cout << std::endl;
    std::cout << "N:     Number of mesh elements, needs to be equal for fluid and structure solver." << std::endl;
    return -1;
  }

  MPI_Init(&argc, &argv);

  int domainSize, gridOffset, rank, size, chunkLength;

  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

  domainSize = atoi(argv[2]);
  if ((domainSize + 1) % size == 0) {
    chunkLength = (domainSize + 1) / size;
    gridOffset = rank * chunkLength;
  } else if (rank < (domainSize + 1) % size) {
    chunkLength = (domainSize + 1) / size + 1;
    gridOffset = rank * chunkLength;
  } else {
    chunkLength = (domainSize + 1) / size;
    gridOffset = ((domainSize + 1) % size) * ((domainSize + 1) / size + 1) + (rank - ((domainSize + 1) % size)) * (domainSize + 1) / size;
  }

  std::vector<double> pressure(chunkLength);
  std::vector<double> crossSectionLength(chunkLength);

  std::string configFileName(argv[1]);
  std::string solverName = "STRUCTURE";

  SolverInterface interface(solverName, configFileName, rank, size);

  int meshID = interface.getMeshID("Structure_Nodes");
  int pressureID = interface.getDataID("Pressure", meshID);
  int crossSectionLengthID = interface.getDataID("CrossSectionLength", meshID);

  int dimensions = interface.getDimensions();
  std::vector<double> grid(dimensions * chunkLength);

  std::vector<int> vertexIDs(chunkLength);

  for (int i = 0; i < chunkLength; i++) {
    crossSectionLength[i] = 1.0;
    pressure[i] = 0.0;
    for (int j = 0; j < dimensions; j++) {
      grid[i * dimensions + j] = j == 0 ? gridOffset + (double)i : 0.0;
    }
  }

  interface.setMeshVertices(meshID, chunkLength, grid.data(), vertexIDs.data());

  interface.initialize();

  double t = 0;
  double dt = 0.01;

  if (interface.isActionRequired(actionWriteInitialData())) {
    interface.writeBlockScalarData(crossSectionLengthID, chunkLength, vertexIDs.data(), crossSectionLength.data());
    interface.fulfilledAction(actionWriteInitialData());
  }

  interface.initializeData();

  if (interface.isReadDataAvailable()) {
    interface.readBlockScalarData(pressureID, chunkLength, vertexIDs.data(), pressure.data());
  }

  while (interface.isCouplingOngoing()) {
    if (interface.isActionRequired(actionWriteIterationCheckpoint())) {
      interface.fulfilledAction(actionWriteIterationCheckpoint());
    }

    structureComputeSolution(rank, size, chunkLength, pressure.data(), crossSectionLength.data()); // Call Solver
                                                                                                   //structureDataDisplay(crossSectionLength, chunkLength);
                                                                                                   //structureDataDisplay(pressure, chunkLength);

    interface.writeBlockScalarData(crossSectionLengthID, chunkLength, vertexIDs.data(), crossSectionLength.data());

    interface.advance(dt);

    interface.readBlockScalarData(pressureID, chunkLength, vertexIDs.data(), pressure.data());

    if (interface.isActionRequired(actionReadIterationCheckpoint())) { // i.e. fluid not yet converged
      interface.fulfilledAction(actionReadIterationCheckpoint());
    } else {
      t += dt;
    }
  }

  interface.finalize();
  MPI_Finalize();

  return 0;
}
